org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/typeHierarchy/TypeHierarchyView.java

```diff
@@ -74,21 +384,95 @@ private void updateHierarchyModeItem(ToolItem hierchyModeItem, boolean showSuper
 				hierchyModeItem.setToolTipText(showSuperTypes ? Messages.typeHierarchy_show_subtypes : Messages.typeHierarchy_show_supertypes);
 			}
 		};
-		treeViewer = filteredTree.getViewer();
-		treeViewer.setContentProvider(contentProvider);
+	}
 
-		treeViewer.setLabelProvider(new TypeHierarchyItemLabelProvider());
+	private SymbolsContainer getSymbolsContainer(IFile file) {
+		return cachedSymbols.computeIfAbsent(file.getLocationURI(), uri -> new SymbolsContainer(file));
+	}
 
-		treeViewer.setUseHashlookup(true);
-		treeViewer.getControl().setEnabled(false);
-		treeViewer.addDoubleClickListener(new IDoubleClickListener() {
-			@Override
-			public void doubleClick(final DoubleClickEvent event) {
-				TypeHierarchyItem item = (TypeHierarchyItem)((IStructuredSelection)event.getSelection()).getFirstElement();
-				LSPEclipseUtils.open(item.getUri(), item.getSelectionRange());
+	private synchronized void refreshSymbols(SymbolsContainer symbolsContainer, boolean documentModified) {
+		if (symbolsContainer == null || (!symbolsContainer.isDirty && !documentModified)) {
+			return;
+		}
+		final IDocument document = symbolsContainer.getDocument();
+		try {
+			if (document != null) {
+				CompletableFuture<List<Either<SymbolInformation, DocumentSymbol>>> symbols;
+				final var params = new DocumentSymbolParams(
+						LSPEclipseUtils.toTextDocumentIdentifier(document));
+				CompletableFuture<Optional<LanguageServerWrapper>> languageServer = LanguageServers
+						.forDocument(document)
+						.withFilter(
+								capabilities -> LSPEclipseUtils.hasCapability(capabilities.getDocumentSymbolProvider()))
```

Comment:
this can just be `.withCapability(ServerCapabilities::getDocumentSymbolProvider)`